### ISSUE TITLE
Contao 3.5.x compatibility with PHP 7.2.x

### ISSUE THREAD

**Issue by moeppe** [contao/core#8837]
PHP now throws an exception when an undefined constant is called, which interrupts much core functionality, like the frontend preview.

For this to solve, every call to 

`if (FE_USER_LOGGED_IN) {`

should be replaced by

`if (defined('FE_USER_LOGGED_IN')) {`

**Comment by aschempp** [comment]
`if (FE_USER_LOGGED_IN) {` is not the same as `if (defined('FE_USER_LOGGED_IN')) {`!
The correct would probably be `if (defined('FE_USER_LOGGED_IN') && true === FE_USER_LOGGED_IN) {`

But why would the constant not be defined in the first place?

**Comment by fritzmg** [comment]
> But why would the constant not be defined in the first place?

This constant is only set in `FrontendIndex::__construct()` (for the front end) and _only_ if the page is not output from the cache. Thus if you have a custom entry point that initializes the Contao framework via the `initialize.php` or if you used an uncached insert tag, then this constant will not be defined by default. 

Related: https://github.com/contao/core-bundle/issues/1244

**Comment by BugBuster1701** [comment]
My Solution therefore is in my __construct():
```
                // See contao/core #4099
		if (!defined('BE_USER_LOGGED_IN'))
		{
			define('BE_USER_LOGGED_IN', false);
		}
		if (!defined('FE_USER_LOGGED_IN'))
		{
			define('FE_USER_LOGGED_IN', false);
		}
```

**Comment by moeppe** [comment]
@aschempp 
You are right. The problem is that (at least my version of) PHP 7.2.1 as FPM service throws an Parse Error, which cannot be prevented. Even in your solution.

would be the following one an solution?:

`if (defined('FE_USER_LOGGED_IN') && true === get_defined_constants()['FE_USER_LOGGED_IN']) {`

For sure it would be much better to ensure the constant is defined in any circumstances.

**Comment by moeppe** [comment]
@BugBuster1701 

With this I'd bre worried about any race conditions.

**Comment by aschempp** [comment]
Why are you using the constant in your code anyway if it might not exist?

**Comment by marcel-mathias-nolte** [comment]
@aschempp 

Would be nice if I were using it. It's everywhere in the core:
https://github.com/contao/core/search?utf8=%E2%9C%93&q=FE_USER_LOGGED_IN&type=

Well my workaround is only a workaround for the symptoms. I have totally no clue about, why this isn't set sometimes. It's definitly never set in Frontend Preview Mode.

PHP Changelog shows that an undefined constant throws an E_WARNING. On my installation it stops the parser. Neither one is good.

**Comment by aschempp** [comment]
I know the constant is used, it is supposed to be. But obviously only if it exists. Can you tell the exact warning where it's used but does not exist?

**Comment by marcel-mathias-nolte** [comment]
e.g

> Warning: Use of undefined constant FE_USER_LOGGED_IN - assumed 'FE_USER_LOGGED_IN' (this will throw an Error in a future version of PHP) in system/modules/core/controllers/BackendSwitch.php

if PHP is not set to strict mode. Otherwise

> Fatal error: Uncaught exception ParseError with message syntax error, unexpected 'FE_USER_LOGGED_IN' (T_STRING) thrown in system/modules/core/controllers/BackendSwitch.php on line 55
> #0 [internal function]: Contao\ClassLoader::load('BackendSwitch')
> #1 contao/switch.php(19): spl_autoload_call('BackendSwitch')
> #2 {main}

which causes the parser to stop. There's no reason to already use PHP 7.2.x exept that it exists .... but it does :-)

**Comment by marcel-mathias-nolte** [comment]
Btw. as I see it, the parse error is caused by the type validation used in strict mode.

Like simple calls to `count($var)` should be replaced by `if ($var instanceof Countable) ...  count($var)` and `if (count($var))` should be replaced by `if (count($var) > 0)` as implicit typecasting is disabled then, which, I think, is a great security feature and makes code much better readable.

**Comment by ausi** [comment]
I was not able to reproduce your syntax error in PHP 7.2 with `strict_types` enabled, see <https://3v4l.org/UR9Jk>

`if (defined('FE_USER_LOGGED_IN') && FE_USER_LOGGED_IN)` should work in every PHP version, regardless of the `strict_types` directive. I think there may be something wrong with your specific PHP version.

Can you create an example code on <https://3v4l.org/> that reproduces the `unexpected T_STRING` syntax error?

**Comment by marcel-mathias-nolte** [comment]
I cannot  reproduce that on that site, maybe it happens due to it's eval'd, maybe something's wrong with my self compiled php version. Even with only the warning frontend preview is unusable, as the control's on the top bar aren't visible because of the warning. Hiding error messages helps using it, but the server log is messed up and it's global. Usally we keep displaying error messages to detect mistakes ... and nearly nobody looks at the error.log in intervals.

We had it at any other place, but I really can't remember.

**Comment by leofeyer** [comment]
The cause of the initial problem in the front end preview was that `contao/switch.php` ran in `FE` mode instead of `BE` mode. I have fixed this in 820847c7c6d88b4dae17fc6c1080353e21466782.

**Comment by k-webdesign** [comment]
Hm... still can reproduce this with PHP 7.2 and Contao 3.5.38 🤷🏼‍♀️
